core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -627,6 +641,29 @@ private[spark] class ExecutorAllocationManager(
     def isExecutorIdle(executorId: String): Boolean = {
       !executorIdToTaskIds.contains(executorId)
     }
+
+    /**
+     * Get the number of locality aware pending tasks and related locality preferences as the
+     * hints used for executor allocation.
+     */
+    def executorPlacementHints(): (Int, Map[String, Int]) =
+      allocationManager.synchronized {
+      var localityAwarePendingTasks: Int = 0
+      val localityToCount = new mutable.HashMap[String, Int]()
+      stageIdToPreferredLocations.values.foreach { localities =>
```

Comment:
Given that the values in stageIdToPreferredLocations don't change after the stage is submitted, it seems like we're going to be redoing a bunch of work each time we sync with the YarnAllocator.  Can we save intermediate information to avoid this?

Also, the way this works, we don't actually differentiate between pending, running, and completed tasks.  All tasks for a stage are counted as pending as long as the stage is still running.  How computationally expensive would it be to keep this up to date when tasks start and complete?